// godspeed-node/dist/core/validation/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.isValidElasticgraphDatasource = exports.isValidKafkaDatasource = exports.isValidRedisDatasource = exports.isValidApiDatasource = exports.isValidWorkflow = exports.isValidEvent = void 0;
const ajv_1 = __importDefault(require("ajv"));
const ajv_formats_1 = __importDefault(require("ajv-formats"));
const event_schema_json_1 = __importDefault(require("./event.schema.json"));
const workflow_schema_json_1 = __importDefault(require("./workflow.schema.json"));
const api_schema_json_1 = __importDefault(require("./datasources/api.schema.json"));
const redis_schema_json_1 = __importDefault(require("./datasources/redis.schema.json"));
const kafka_schema_json_1 = __importDefault(require("./datasources/kafka.schema.json"));
const elasticgraph_schema_json_1 = __importDefault(require("./datasources/elasticgraph.schema.json"));
const logger_1 = require("../../logger");
const ajvInstance = new ajv_1.default({ allErrors: true, coerceTypes: true });
(0, ajv_formats_1.default)(ajvInstance);
require('ajv-errors')(ajvInstance);
const validateEvent = ajvInstance.compile(event_schema_json_1.default);
const validateWorkflow = ajvInstance.compile(workflow_schema_json_1.default);
const validateApiDs = ajvInstance.compile(api_schema_json_1.default);
const validateRedisDs = ajvInstance.compile(redis_schema_json_1.default);
const validateKafkaDs = ajvInstance.compile(kafka_schema_json_1.default);
const validElasticgraphDs = ajvInstance.compile(elasticgraph_schema_json_1.default);
const isValidEvent = (event, eventKey) => {
    if (!validateEvent(event)) {
        logger_1.logger.error('Event validation failed for %s', eventKey);
        logger_1.logger.error(validateEvent.errors);
        return false;
    }
    return true;
};
exports.isValidEvent = isValidEvent;
const isValidWorkflow = (workflow, workflowKey) => {
    if (!validateWorkflow(workflow)) {
        logger_1.logger.error('Workflow validation failed for %s', workflowKey);
        logger_1.logger.error(validateWorkflow.errors);
        return false;
    }
    return true;
};
exports.isValidWorkflow = isValidWorkflow;
const isValidApiDatasource = (datasource) => {
    if (!validateApiDs(datasource)) {
        logger_1.logger.error('Datasource validation failed for %s datasource', JSON.stringify(datasource));
        logger_1.logger.error(validateApiDs.errors);
        return false;
    }
    return true;
};
exports.isValidApiDatasource = isValidApiDatasource;
const isValidRedisDatasource = (datasource) => {
    if (!validateRedisDs(datasource)) {
        logger_1.logger.error('Datasource validation failed for %s datasource', JSON.stringify(datasource));
        logger_1.logger.error(validateRedisDs.errors);
        return false;
    }
    return true;
};
exports.isValidRedisDatasource = isValidRedisDatasource;
const isValidKafkaDatasource = (datasource) => {
    if (!validateKafkaDs(datasource)) {
        logger_1.logger.error('Datasource validation failed for %s datasource', JSON.stringify(datasource));
        logger_1.logger.error(validateKafkaDs.errors);
        return false;
    }
    return true;
};
exports.isValidKafkaDatasource = isValidKafkaDatasource;
const isValidElasticgraphDatasource = (datasource) => {
    if (!validElasticgraphDs(datasource)) {
        logger_1.logger.error('Datasource validation failed for %s datasource', JSON.stringify(datasource));
        return false;
    }
    return true;
};
exports.isValidElasticgraphDatasource = isValidElasticgraphDatasource;
exports.default = ajvInstance;
